### PR TITLE
flamenco: loader v4

### PR DESCRIPTION
**Pull request by mjain-jump**

### CODE REVIEW

File: src/flamenco/runtime/program/fd_loader_v4_program.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+          or already deployed and in maintenance (and thus cannot be invoked).
+      - Deployed
+        - This is the normal state for a program that is ready to be invoked.
+        - Programs cannot be retracted within `LOADER_V4_DEPLOYMENT_COOLDOWN_IN_SLOTS` (750) slots of deployment.
+      - Finalized
+        - The program is immutable.
+        - Users must specify a "next version" which, from my inspection, serves no functional purpose besides showing up
+          as extra information on a block explorer.
+    - There is no longer a concept of a program account vs. a program data account. The program account is the program data account.
+      - "Look at me... I'm the programdata account now..."
```
Comment (ravyu-jump) [inline src/flamenco/runtime/program/fd_loader_v4_program.h:26]: nit: remove

**Review by topointon-jump — COMMENTED** [review]
This is awesome! Only some minor comments.

Also, are there any ledger/fuzz tests for this?

File: src/flamenco/runtime/program/fd_bpf_loader_program.h [hunk 2]
```diff
@@ -54,6 +55,14 @@ fd_bpf_loader_v3_program_get_state( fd_exec_instr_ctx_t *                instr_c
...
                                      fd_bpf_upgradeable_loader_state_t * state );
 
+int
+fd_deploy_program( fd_exec_instr_ctx_t * instr_ctx,
+                   uchar const *         programdata,
+                   ulong                 programdata_size );
+
+int
+fd_bpf_execute( fd_exec_instr_ctx_t * instr_ctx, fd_sbpf_validated_program_t * prog, uchar is_deprecated );
+
```
[the remark this answers is not in the document]
Comment (mjain-jump) [reply src/flamenco/runtime/program/fd_bpf_loader_program.h:65]: The .c file has the comments for these functions

File: src/flamenco/types/gen_stubs.py [hunk 3]
```diff
@@ -1989,6 +1989,15 @@ def __init__(self, json):
         self.compact = (json["compact"] if "compact" in json else False)
         self.archival = (bool(json["archival"]) if "archival" in json else False)
 
+        # Current supported repr types for enum are uint and ulong
```
[the remark this answers is not in the document]
Comment (mjain-jump) [reply src/flamenco/types/gen_stubs.py:1992]: It is but it still deserializes as a uint, I think this is some underlying behavior with how bincode deserialization works. [Take this devnet instruction for example](https://explorer.solana.com/tx/f3q7DptyMPMDKuH5geJVDvxuz4afYcLpCjWVCCJHwVBoqxjvkFEMzp9VmBbYBiNLg82dCniEAebdXNn3G9aPAeW?cluster=devnet), the first 4 bytes are the discriminant for the write instruction.

**Comment by mjain-jump** [comment]
Agave's implementation on their latest release is incomplete. For now I've been fuzz-testing my implementation locally on my box. Once their implementation is complete and cut into a release I'll add fuzz and ledger tests.

**Review by topointon-jump — APPROVED** [review]
🔥

**Pull request merged**